File: Day01/golang/program.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"strconv"
	"strings"
)

const (
	North int = iota
	East
	South
	West
)

type location struct {
	x int
	y int
}

func run() int {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "%s filename\n", os.Args[0])
		return 1
	}

	b, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	movements := make(map[int]int, 0)
	visited := make([]location, 0)

	currentDirection := North

	pos := location{0, 0}

	instructions := strings.Split(string(b), ", ")

	for _, instruction := range instructions {

		turn := int(instruction[0])
		distance, _ := strconv.Atoi(instruction[1:])

		if turn == 'R' {
			currentDirection++
		} else {
			currentDirection--
		}

		// TODO: Investigate if Python-like enums could be used such this if can be avoided
		if currentDirection < 0 {
			currentDirection = 3
		}
		currentDirection %= 4

		for i := 0; i < distance; i++ {

			movements[currentDirection]++

			pos = location{movements[East] - movements[West], movements[North] - movements[South]}

			if Any(visited, pos) {
				blocks := math.Abs(float64(pos.x)) + math.Abs(float64(pos.y))
				fmt.Println("Location is already visited @ (", pos.x, ",", pos.y, ") => ", blocks, " blocks away")
			} else {
				visited = append(visited, pos)
			}
		}
	}

	blocks := math.Abs(float64(pos.x)) + math.Abs(float64(pos.y))
	fmt.Println("Final location is @ (", pos.x, ",", pos.y, ") => ", blocks, " blocks away")

	return 0
}

func Any(all []location, target location) bool {
	for _, pos := range all {
		if pos.x == target.x && pos.y == target.y {
			return true
		}
	}
	return false
}

func main() {
	fmt.Println("+-------------------------+")
	fmt.Println("| Advent of Code - Day 01 |")
	fmt.Println("+-------------------------+")

	result := run()
	os.Exit(result)
}

File: Day01/DotNet/Program.cs
using System;
using System.IO;
using System.Linq;
using System.Collections.Generic;

namespace Day1.ConsoleApplication
{
    public class Program
    {
        // The %-operator of C# is actually NOT a modulo for reasons?! 
        private static int mod(int k, int n) {  return ((k %= n) < 0) ? k+n : k;  }

        public static void Main(string[] args)
        {
            Console.WriteLine("+-------------------------+");
            Console.WriteLine("| Advent of Code - Day 01 |");
            Console.WriteLine("+-------------------------+");
            Console.WriteLine();

            var input = File.ReadAllText("input.txt")
                .Split(new []{','}, StringSplitOptions.RemoveEmptyEntries)
                .Select(e => e.Trim());

            var position = new Tuple<int,int>(0, 0);
            var visits = new List<Tuple<int,int>> { position };
            var moves = new int[4];

            var currentDirection = 0; // North=0, 1=East, 2=South, 3=West 

            foreach(var instruction in input)
            {
                var turn = instruction[0];
                int length = int.Parse(new string(instruction.Skip(1).ToArray()));
                
                if(turn == 'R')
                    ++currentDirection;
                else
                    --currentDirection;

                currentDirection = mod(currentDirection, 4);

                for(int i=0; i<length; ++i)
                {
                    moves[currentDirection]++;
                    position = new Tuple<int, int>(moves[1] - moves[3], moves[0] - moves[2]);

                    if(visits.Any(x => x.Item1 == position.Item1 && x.Item2 == position.Item2))
                    {
                        int firstBlocks = Math.Abs(position.Item1) + Math.Abs(position.Item2);
                        Console.WriteLine($"Location is already visited @ ({position.Item1}, {position.Item2}) => {firstBlocks} blocks away");
                    } else {
                        visits.Add(position);
                    }
                }
            }
            int finalBlocks = Math.Abs(position.Item1) + Math.Abs(position.Item2);
            Console.WriteLine($"Final location is @ ({position.Item1},{position.Item2} blocks {finalBlocks} away");
            Console.WriteLine($"  -Gruss Gott");
        }
    }
}